iml-agent/src/network.rs

```diff
@@ -0,0 +1,449 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use combine::{
+    attempt, choice,
+    error::ParseError,
+    many1, optional,
+    parser::{
+        char::{char, digit, letter, spaces, string},
+        repeat::take_until,
+    },
+    sep_by1,
+    stream::Stream,
+    token, Parser,
+};
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum InterfaceProperties {
+    InterfaceFlagsAndAttributes((String, Vec<String>, HashMap<String, String>)),
+    InterfaceTypeAndMacAddress((Option<String>, Option<String>)),
+    Inet4AddressAndPrefix((String, u32)),
+    Inet6AddressAndPrefix((String, u32)),
+}
+
+#[derive(Debug)]
+pub struct Address {
+    address: String,
+    prefix: u32,
+}
+
+#[derive(Debug)]
+pub struct NetworkInterface {
+    pub interface: String,
+    pub mac_address: Option<String>,
+    pub interface_type: Option<String>,
+    pub inet4_address: Vec<Address>,
+    pub inet6_address: Vec<Address>,
+    pub is_up: bool,
+    pub is_slave: bool,
+}
+
+impl NetworkInterface {
+    pub fn set_prop(mut self, prop: InterfaceProperties) -> NetworkInterface {
+        match prop {
+            InterfaceProperties::InterfaceFlagsAndAttributes((interface, flags, _attributes)) => {
+                self.interface = interface;
+
+                if flags.iter().any(|x| x == &"UP".to_string()) {
+                    self.is_up = true;
+                }
+
+                if flags.iter().any(|x| x == &"SLAVE".to_string()) {
+                    self.is_slave = true;
+                }
+            }
+            InterfaceProperties::InterfaceTypeAndMacAddress((interface_type, address)) => {
+                self.interface_type = interface_type;
+                self.mac_address = address;
+            }
+            InterfaceProperties::Inet4AddressAndPrefix((address, prefix)) => {
+                self.inet4_address.push(Address { address, prefix });
+            }
+            InterfaceProperties::Inet6AddressAndPrefix((address, prefix)) => {
+                self.inet6_address.push(Address { address, prefix });
+            }
+        }
+
+        self
+    }
+}
+
+impl Default for NetworkInterface {
```

Comment:
Why not just derive this?

Reply:
By derive, are you suggesting that I get rid of the default implementation and just create the struct since I'm only using it in one place?

Reply:
No, I'm suggesting you use `#[derive(Default)]` instead of writing it by hand.